apps/src/templates/sectionProgressV2/floatingScrollbar/FloatingScrollbar.jsx

```diff
@@ -0,0 +1,142 @@
+ * childContentsRef - a ref to the contents of {child} element.
+ *    Set by calling `ref={childContentsRef}` on the contents of {child} element
+ */
+export default function FloatingScrollbar({
+  children,
+  setOnScroll,
+  childContainerRef,
+  childContentsRef,
+}) {
+  const scrollRef = React.useRef();
+
+  const [childScrollWidth, setChildScrollWidth] = React.useState(0);
+  const [childWidth, setChildWidth] = React.useState(0);
+  const [scrollVisible, setScrollVisible] = React.useState(true);
+
+  if (children.length > 1) {
+    throw new Error('FloatingScrollbar only supports a single child');
+  }
+
+  const childContainerResizeObserver = React.useMemo(
+    () =>
+      new ResizeObserver(([entry]) => {
+        if (entry.borderBoxSize) {
+          setChildWidth(entry.borderBoxSize[0].inlineSize);
+        }
+      }),
+    [setChildWidth]
+  );
+
+  const childContentsResizeObserver = React.useMemo(
+    () =>
+      new ResizeObserver(([entry]) => {
+        if (entry.borderBoxSize) {
+          setChildScrollWidth(entry.borderBoxSize[0].inlineSize + 1);
+        }
+      }),
+    [setChildScrollWidth]
+  );
+
+  React.useEffect(() => {
+    if (childContainerResizeObserver) {
+      childContainerResizeObserver.observe(childContainerRef.current);
+    }
+    if (childContentsResizeObserver) {
+      childContentsResizeObserver.observe(childContentsRef.current);
+    }
+    return () => {
+      childContainerResizeObserver.disconnect();
+      childContentsResizeObserver.disconnect();
+    };
+  }, [
+    childContainerResizeObserver,
+    childContentsResizeObserver,
+    childContainerRef,
+    childContentsRef,
+  ]);
+
+  const handleScrollAndResize = React.useCallback(() => {
+    const maxVisibleY =
+      window.innerHeight || document.documentElement.clientHeight;
+
+    // 20 is the height of the scrollbar.
+    const isNowVisible =
+      childContainerRef?.current.getBoundingClientRect().bottom + 20 <
+      maxVisibleY;
+
+    if (isNowVisible !== scrollVisible) {
+      setScrollVisible(isNowVisible);
+    }
+  }, [childContainerRef, scrollVisible, setScrollVisible]);
+
+  React.useEffect(() => {
+    window.addEventListener('scroll', e => handleScrollAndResize(e));
+    window.addEventListener('resize', e => handleScrollAndResize(e));
+
+    return () => {
+      // return a cleanup function to unregister our function since it will run multiple times
+      window.removeEventListener('scroll', e => handleScrollAndResize(e));
+      window.removeEventListener('resize', e => handleScrollAndResize(e));
+    };
+  }, [handleScrollAndResize]);
+
+  const scrollChild = React.useCallback(
+    scroll => {
+      if (childContainerRef?.current) {
+        childContainerRef.current.scrollLeft = scroll.target.scrollLeft;
+      }
+    },
+    [childContainerRef]
+  );
+
+  React.useEffect(() => {
+    setOnScroll(() => scroll => {
+      scrollRef.current.scrollLeft = scroll.target.scrollLeft;
```

Comment:
Have you tried this in RTL mode?

Reply:
`scrollLeft` is the keyword for horizontal scroll. There is no `scrollRight`.

Reply:
Yep. I was reading the [documentation about this](https://developer.mozilla.org/en-US/docs/Web/API/Element/scrollLeft) and it says

> If the element's [direction](https://developer.mozilla.org/en-US/docs/Web/CSS/direction) is rtl (right-to-left), then scrollLeft is 0 when the scrollbar is at its rightmost position (at the start of the scrolled content), and then increasingly negative as you scroll towards the end of the content.

Seems worth a test.